Make ROOT file support concurrent LuminosityBlocks

Dr15Jones opened the pull request:
The PoolOutputModule can now write files where an Event from the next LuminosityBlock is written to the output file before the end LuminosityBlock transition occurs. It can even handle interleaving Events from two different LuminosityBlocks before the previous LuminosityBlock ends.

The PoolSource can read files containing such interleaved Events. When reading back, it returns all the Events for one LuminosityBlock before going onto the next LuminosityBlock. [We already did such behavior in the cases where the output file was created by a job which read the LuminosityBlock from different files].

Dr15Jones committed:
IndexIntoFile now handles overlapping LuminosityBlocks

IndexIntoFile was modified to handle the case where we are processing LuminosityBlocks concurrently and events from the next LuminosityBlock are being written before the endLuminosityBlock transition from the previous LuminosityBlock has been called.
This was accomplished by adding a Lumi entry to runOrLumiEntries_ with an invalidEntry value which is there to keep track of the event entry range for the already written events for the previous LuminosityBlock.

Dr15Jones committed:
Handle unordered transitions in RunLumiEventChecker

RunLumiEventChecker can now be used in multi-threaded jobs (or to read files created by multi-threaded jobs) where the order of the transitions can change process to process.
Also made RunLumiEventChecker a global module.

Dr15Jones committed:
Unit test for overlapping Lumi write/read

With the use of concurrent LuminosityBlocks, it is possible for an event of the next LuminosityBlock to be written to the ROOT file before the end LuminosityBlock transition has been seen by the OutputModule. This test checks that PoolOutputModule can properly write files in that case and that PoolSource can subsequently read them.

Dr15Jones commented:
@wddgit please take a look

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-22406/3646

cmsbuild commented:
A new Pull Request was created by  @Dr15Jones (Chris Jones) for master.

It involves the following packages:

DataFormats/Provenance
FWCore/Modules
IOPool/Input


@cmsbuild, @smuzaffar, @Dr15Jones can you please review it and eventually sign? Thanks.
@makortel, @Martin-Grunewald, @rovere, @wddgit this is something you requested to watch as well.
@davidlange6, @slava77, @fabiocos you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

Dr15Jones commented:
@dan131riley does this modification also make it easier for you to implement the limited PoolOutputModule?

Dr15Jones commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: f886db9eb55006ab86cd081308889b98c554350b
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-22406/26415/summary.html

The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/677ceb2e382b709b70741d68c0b3515361c9a7b4
https://github.com/cms-sw/cmssw/commit/14174f204ded792cded2463134812d06a19f962a
https://github.com/cms-sw/cmssw/commit/98262121c7b408a1379be7f479933f91c52c534d
https://github.com/cms-sw/cmssw/commit/2a879bdb5c4d181d425af6d3c03cb6c37653e328
https://github.com/cms-sw/cmssw/commit/60b192bb2d3ce2dbb7cd6785291b87a18fa35692
https://github.com/cms-sw/cmssw/commit/c1ce9e55401411995e0edef25f111bd3b4d34356
https://github.com/cms-sw/cmssw/commit/03490e6b5bbb0d4a78dbdf7572422ec60767778d
https://github.com/cms-sw/cmssw/commit/bd14132d32453f5c01603a221955d980d41a62f2
https://github.com/cms-sw/cmssw/commit/783236f35c3adf29544ec06fc11d1b359e7c1731
https://github.com/cms-sw/cmssw/commit/236ebf72092fa8158d3364f77aa17ddbad7ee85f
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-22406/26415/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-22406/26415/git-merge-result

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-22406/26415/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 1009 differences found in the comparisons
 - DQMHistoTests: Total files compared: 29
 - DQMHistoTests: Total histograms compared: 2479021
 - DQMHistoTests: Total failures: 4692
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2474153
 - DQMHistoTests: Total skipped: 176
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 1.11 KiB( 21 files compared)
 - Checked 118 log files, 9 edm output root files, 29 DQM output files

Dr15Jones commented:
The differences with the base line are most likely caused by recent changes in the IB (since they also show up in other pull requests). Just to be safe, I'll rerun the tests

Dr15Jones commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: f886db9eb55006ab86cd081308889b98c554350b
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-22406/26444/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-22406/26444/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 29
 - DQMHistoTests: Total histograms compared: 2479021
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2478844
 - DQMHistoTests: Total skipped: 176
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 1.24000000005 KiB( 23 files compared)
 - Checked 118 log files, 9 edm output root files, 29 DQM output files

wddgit commented:
In previousLumiWithEvents, after it finds a new lumi it needs to iterate forward across the lumi entries with invalid TTree entry numbers. I'm still reading through this to look for problems. I'll let you know when I am done. Not sure if you want me to actually fix things or just comments like above ... Can I submit a pull request to a pull request or edit a pull request?

wddgit commented:
There are two places in the function findPosition(RunNumber_t run, LuminosityBlockNumber_t lumi, EventNumber_t event) where the IndexIntoFileItr constructor is explicitly called specifying indexToLumi. In both, that index needs to step over the invalid values.

Dr15Jones commented:
Could you send me a unit test exhibiting the problem? A fix would be great.

wddgit committed:
More, make IndexIntoFile work with concurrent lumis

Earlier commits got most of it working,
but this gets the sorted iteration working,
the findPosition functions, and backwards
skipping. Also modified comments and variable
names where appropriate and extended the unit
test.

wddgit commented:
@Dr15Jones 
I made the changes to fix the problems I noticed and also
added unit tests to cover those cases. I also modified some
comments and variable names where appropriate. I tried to
push to your branch but do not have the privileges needed
to do that, so you can pull the branch I created from here:
https://github.com/wddgit/cmssw/commits/concurrentLumisIOHandling
There is one commit on top of your commits.

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-22406/3824

cmsbuild commented:
Pull request #22406 was updated. @cmsbuild, @smuzaffar, @Dr15Jones can you please check and sign again.

Dr15Jones commented:
please test

cmsbuild commented:
+1
Tested at: 9e001416522cf050d512f0adc568b6e62f7f4a95
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-22406/26677/summary.html

cmsbuild commented:
Comparison job queued.

Dr15Jones commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @davidlange6, @slava77, @smuzaffar, @fabiocos (and backports should be raised in the release meeting by the corresponding L2)

fabiocos commented:
+1
